File: bindings/julia/test/test_lua_runner.jl
module TestLuaRunner

using QUIVERDatabase
using Test

include("fixture.jl")

@testset "LuaRunner" begin
    @testset "Create Element" begin
        path_schema = joinpath(tests_path(), "schemas", "valid", "collections.sql")
        db = QUIVERDatabase.from_schema(":memory:", path_schema)

        lua = QUIVERDatabase.LuaRunner(db)

        QUIVERDatabase.run!(
            lua,
            """
        db:create_element("Configuration", { label = "Test Config" })
        db:create_element("Collection", { label = "Item 1", some_integer = 42 })
    """,
        )

        labels = QUIVERDatabase.read_scalar_strings(db, "Collection", "label")
        @test length(labels) == 1
        @test labels[1] == "Item 1"

        integers = QUIVERDatabase.read_scalar_integers(db, "Collection", "some_integer")
        @test length(integers) == 1
        @test integers[1] == 42

        QUIVERDatabase.close!(lua)
        QUIVERDatabase.close!(db)
    end

    @testset "Read from Lua" begin
        path_schema = joinpath(tests_path(), "schemas", "valid", "collections.sql")
        db = QUIVERDatabase.from_schema(":memory:", path_schema)

        QUIVERDatabase.create_element!(db, "Configuration"; label = "Config")
        QUIVERDatabase.create_element!(db, "Collection"; label = "Item 1", some_integer = 10)
        QUIVERDatabase.create_element!(db, "Collection"; label = "Item 2", some_integer = 20)

        lua = QUIVERDatabase.LuaRunner(db)

        QUIVERDatabase.run!(
            lua,
            """
        local labels = db:read_scalar_strings("Collection", "label")
        assert(#labels == 2, "Expected 2 labels")
        assert(labels[1] == "Item 1", "First label mismatch")
        assert(labels[2] == "Item 2", "Second label mismatch")
    """,
        )

        QUIVERDatabase.close!(lua)
        QUIVERDatabase.close!(db)
    end

    @testset "Script Error" begin
        path_schema = joinpath(tests_path(), "schemas", "valid", "collections.sql")
        db = QUIVERDatabase.from_schema(":memory:", path_schema)

        lua = QUIVERDatabase.LuaRunner(db)

        @test_throws QUIVERDatabase.DatabaseException QUIVERDatabase.run!(lua, "invalid syntax !!!")

        QUIVERDatabase.close!(lua)
        QUIVERDatabase.close!(db)
    end

    @testset "Reuse Runner" begin
        path_schema = joinpath(tests_path(), "schemas", "valid", "collections.sql")
        db = QUIVERDatabase.from_schema(":memory:", path_schema)

        lua = QUIVERDatabase.LuaRunner(db)

        QUIVERDatabase.run!(lua, """db:create_element("Configuration", { label = "Config" })""")
        QUIVERDatabase.run!(lua, """db:create_element("Collection", { label = "Item 1" })""")
        QUIVERDatabase.run!(lua, """db:create_element("Collection", { label = "Item 2" })""")

        labels = QUIVERDatabase.read_scalar_strings(db, "Collection", "label")
        @test length(labels) == 2
        @test labels[1] == "Item 1"
        @test labels[2] == "Item 2"

        QUIVERDatabase.close!(lua)
        QUIVERDatabase.close!(db)
    end

    # Error handling tests

    @testset "Undefined Variable" begin
        path_schema = joinpath(tests_path(), "schemas", "valid", "collections.sql")
        db = QUIVERDatabase.from_schema(":memory:", path_schema)

        lua = QUIVERDatabase.LuaRunner(db)

        # Script that references undefined variable
        @test_throws QUIVERDatabase.DatabaseException QUIVERDatabase.run!(lua, "print(undefined_variable.field)")

        QUIVERDatabase.close!(lua)
        QUIVERDatabase.close!(db)
    end

    @testset "Create Invalid Collection" begin
        path_schema = joinpath(tests_path(), "schemas", "valid", "collections.sql")
        db = QUIVERDatabase.from_schema(":memory:", path_schema)

        lua = QUIVERDatabase.LuaRunner(db)

        QUIVERDatabase.run!(lua, """db:create_element("Configuration", { label = "Test Config" })""")

        # Script that creates element in nonexistent collection
        @test_throws QUIVERDatabase.DatabaseException QUIVERDatabase.run!(
            lua,
            """db:create_element("NonexistentCollection", { label = "Item" })""",
        )

        QUIVERDatabase.close!(lua)
        QUIVERDatabase.close!(db)
    end

    @testset "Empty Script" begin
        path_schema = joinpath(tests_path(), "schemas", "valid", "collections.sql")
        db = QUIVERDatabase.from_schema(":memory:", path_schema)

        lua = QUIVERDatabase.LuaRunner(db)

        # Empty script should succeed without error
        QUIVERDatabase.run!(lua, "")
        @test true  # If we get here, the empty script ran without error

        QUIVERDatabase.close!(lua)
        QUIVERDatabase.close!(db)
    end

    @testset "Comment Only Script" begin
        path_schema = joinpath(tests_path(), "schemas", "valid", "collections.sql")
        db = QUIVERDatabase.from_schema(":memory:", path_schema)

        lua = QUIVERDatabase.LuaRunner(db)

        # Comment-only script should succeed
        QUIVERDatabase.run!(lua, "-- this is just a comment")
        @test true

        QUIVERDatabase.close!(lua)
        QUIVERDatabase.close!(db)
    end

    @testset "Read Integers" begin
        path_schema = joinpath(tests_path(), "schemas", "valid", "collections.sql")
        db = QUIVERDatabase.from_schema(":memory:", path_schema)

        QUIVERDatabase.create_element!(db, "Configuration"; label = "Config")
        QUIVERDatabase.create_element!(db, "Collection"; label = "Item 1", some_integer = 100)
        QUIVERDatabase.create_element!(db, "Collection"; label = "Item 2", some_integer = 200)

        lua = QUIVERDatabase.LuaRunner(db)

        QUIVERDatabase.run!(
            lua,
            """
        local ints = db:read_scalar_integers("Collection", "some_integer")
        assert(#ints == 2, "Expected 2 integers")
        assert(ints[1] == 100, "First integer mismatch")
        assert(ints[2] == 200, "Second integer mismatch")
    """,
        )

        QUIVERDatabase.close!(lua)
        QUIVERDatabase.close!(db)
    end

    @testset "Read Floats" begin
        path_schema = joinpath(tests_path(), "schemas", "valid", "collections.sql")
        db = QUIVERDatabase.from_schema(":memory:", path_schema)

        QUIVERDatabase.create_element!(db, "Configuration"; label = "Config")
        QUIVERDatabase.create_element!(db, "Collection"; label = "Item 1", some_float = 1.5)
        QUIVERDatabase.create_element!(db, "Collection"; label = "Item 2", some_float = 2.5)

        lua = QUIVERDatabase.LuaRunner(db)

        QUIVERDatabase.run!(
            lua,
            """
        local floats = db:read_scalar_floats("Collection", "some_float")
        assert(#floats == 2, "Expected 2 floats")
        assert(floats[1] == 1.5, "First float mismatch")
        assert(floats[2] == 2.5, "Second float mismatch")
    """,
        )

        QUIVERDatabase.close!(lua)
        QUIVERDatabase.close!(db)
    end

    @testset "Read Vectors" begin
        path_schema = joinpath(tests_path(), "schemas", "valid", "collections.sql")
        db = QUIVERDatabase.from_schema(":memory:", path_schema)

        QUIVERDatabase.create_element!(db, "Configuration"; label = "Config")
        QUIVERDatabase.create_element!(db, "Collection"; label = "Item 1", value_int = [1, 2, 3])

        lua = QUIVERDatabase.LuaRunner(db)

        QUIVERDatabase.run!(
            lua,
            """
        local vectors = db:read_vector_integers("Collection", "value_int")
        assert(#vectors == 1, "Expected 1 vector")
        assert(#vectors[1] == 3, "Expected 3 elements in vector")
        assert(vectors[1][1] == 1, "First element mismatch")
        assert(vectors[1][2] == 2, "Second element mismatch")
        assert(vectors[1][3] == 3, "Third element mismatch")
    """,
        )

        QUIVERDatabase.close!(lua)
        QUIVERDatabase.close!(db)
    end

    @testset "Create With Vector" begin
        path_schema = joinpath(tests_path(), "schemas", "valid", "collections.sql")
        db = QUIVERDatabase.from_schema(":memory:", path_schema)

        lua = QUIVERDatabase.LuaRunner(db)

        QUIVERDatabase.run!(
            lua,
            """
        db:create_element("Configuration", { label = "Config" })
        db:create_element("Collection", { label = "Item 1", value_int = {10, 20, 30} })
    """,
        )

        result = QUIVERDatabase.read_vector_integers(db, "Collection", "value_int")
        @test length(result) == 1
        @test result[1] == [10, 20, 30]

        QUIVERDatabase.close!(lua)
        QUIVERDatabase.close!(db)
    end
end

end

File: bindings/python/tests/test_lua_runner.py
from __future__ import annotations

import json

import pytest

from quiverdb import Database, LuaRunner, QuiverError


class TestLuaRunnerCreateRead:
    """Tests for Lua scripts that create and read elements."""

    def test_create_element_from_lua(self, collections_db: Database) -> None:
        lua = LuaRunner(collections_db)
        lua.run("""
            db:create_element("Configuration", { label = "default" })
            db:create_element("Collection", { label = "Item1", some_integer = 42, some_float = 3.14 })
        """)
        labels = collections_db.read_scalar_strings("Collection", "label")
        assert labels == ["Item1"]
        values = collections_db.read_scalar_integers("Collection", "some_integer")
        assert values == [42]
        lua.close()

    def test_read_scalars_from_lua(self, collections_db: Database) -> None:
        collections_db.create_element("Configuration", label="default")
        collections_db.create_element("Collection", label="Seeded", some_integer=99)
        lua = LuaRunner(collections_db)
        lua.run("""
            local labels = db:read_scalar_strings("Collection", "label")
            assert(#labels == 1, "Expected 1 label, got " .. #labels)
            assert(labels[1] == "Seeded", "Expected 'Seeded', got " .. labels[1])
        """)
        lua.close()


class TestLuaRunnerErrors:
    """Tests for Lua script error handling."""

    def test_syntax_error_raises_quiver_error(self, collections_db: Database) -> None:
        lua = LuaRunner(collections_db)
        with pytest.raises(QuiverError) as exc_info:
            lua.run("invalid syntax !!!")
        assert str(exc_info.value) != ""
        lua.close()

    def test_runtime_error_raises_quiver_error(self, collections_db: Database) -> None:
        lua = LuaRunner(collections_db)
        with pytest.raises(QuiverError):
            lua.run("print(undefined_variable.field)")
        lua.close()

    def test_invalid_collection_raises_quiver_error(self, collections_db: Database) -> None:
        collections_db.create_element("Configuration", label="default")
        lua = LuaRunner(collections_db)
        with pytest.raises(QuiverError):
            lua.run("""
                db:create_element("NonexistentCollection", { label = "Bad" })
            """)
        lua.close()


class TestLuaRunnerLifecycle:
    """Tests for LuaRunner lifecycle management."""

    def test_multiple_run_calls(self, collections_db: Database) -> None:
        lua = LuaRunner(collections_db)
        lua.run('db:create_element("Configuration", { label = "default" })')
        lua.run('db:create_element("Collection", { label = "Item1", some_integer = 1 })')
        lua.run('db:create_element("Collection", { label = "Item2", some_integer = 2 })')
        labels = collections_db.read_scalar_strings("Collection", "label")
        assert sorted(labels) == ["Item1", "Item2"]
        lua.close()

    def test_empty_script(self, collections_db: Database) -> None:
        lua = LuaRunner(collections_db)
        lua.run("")
        lua.close()

    def test_comment_only_script(self, collections_db: Database) -> None:
        lua = LuaRunner(collections_db)
        lua.run("-- just a comment")
        lua.close()

    def test_context_manager(self, collections_db: Database) -> None:
        with LuaRunner(collections_db) as lua:
            lua.run('db:create_element("Configuration", { label = "default" })')
            labels = collections_db.read_scalar_strings("Configuration", "label")
            assert labels == ["default"]
        with pytest.raises(QuiverError, match="LuaRunner is closed"):
            lua.run("-- should fail")

    def test_run_after_close_raises(self, collections_db: Database) -> None:
        lua = LuaRunner(collections_db)
        lua.close()
        with pytest.raises(QuiverError, match="LuaRunner is closed"):
            lua.run("-- should fail")

    def test_close_idempotent(self, collections_db: Database) -> None:
        lua = LuaRunner(collections_db)
        lua.close()
        lua.close()  # Should not raise

    def test_database_reference_kept(self, collections_db: Database) -> None:
        lua = LuaRunner(collections_db)
        assert lua._db is collections_db
        lua.close()


class TestLuaRunnerReturnValues:
    """A script hands one value back to the caller as JSON."""

    def test_returns_json(self, collections_db: Database) -> None:
        lua = LuaRunner(collections_db)
        assert lua.run("return { a = 1, b = { 2, 3 } }") == '{"a":1,"b":[2,3]}'
        assert json.loads(lua.run('return db:read_element_ids("Collection")')) == []

    def test_returns_empty_string_when_script_returns_nothing(self, collections_db: Database) -> None:
        lua = LuaRunner(collections_db)
        assert lua.run("local x = 1") == ""


class TestDatabaseDryRun:
    """A dry run executes writes and throws them away."""

    def test_rolls_back_a_script(self, collections_db: Database) -> None:
        lua = LuaRunner(collections_db)
        lua.run('db:create_element("Configuration", { label = "default" })')

        assert collections_db.in_dry_run() is False
        with collections_db.dry_run():
            assert collections_db.in_dry_run() is True
            # db:transaction composes: the dry run absorbs the nested BEGIN/COMMIT.
            result = lua.run("""
                db:transaction(function(db)
                    db:create_element("Collection", { label = "Preview" })
                end)
                return db:read_scalar_strings("Collection", "label")
            """)
            assert json.loads(result) == ["Preview"]

        assert collections_db.in_dry_run() is False
        assert collections_db.read_scalar_strings("Collection", "label") == []

    def test_rolls_back_on_exception(self, collections_db: Database) -> None:
        lua = LuaRunner(collections_db)
        lua.run('db:create_element("Configuration", { label = "default" })')

        with pytest.raises(ValueError):
            with collections_db.dry_run():
                lua.run('db:create_element("Collection", { label = "Preview" })')
                raise ValueError("boom")

        assert collections_db.read_scalar_strings("Collection", "label") == []
        assert collections_db.in_dry_run() is False

    def test_end_without_start_raises(self, collections_db: Database) -> None:
        with pytest.raises(QuiverError, match="no active dry run"):
            collections_db.end_dry_run()
